### PR TITLE
fix(#702): sweep-invariant 2x2 CTM corner label convention

### LINKED ISSUE #702: Single-site 2x2 CTM chi-bump: forward fixed point trajectory-dependent + implicit-AD gradient ~10x too small — regression from PR #676
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

## Summary

The single-site **implicit-AD chi-bump** path (trivial-U(1) `DenseTensor`, 2x2 plaquette projector) regressed in two ways, both bisected to a **clean first-bad commit**: `7b8e5ad` — **PR #676** (`fix(#670): symmetric 2x2 CTM direction-dependent bond bookkeeping (unblocks #667)`).

1. **Forward fixed point is now trajectory-dependent.** The chi 4→8 bump path and the direct-chi=8 path no longer converge to the same CTM fixed point (they must — the fixed point is unique).
2. **Implicit-AD gradient is ~10× too small** vs central-difference on the bump path (27 of 32 indices fail the tolerance) — consistent with the chi-lock backward differentiating the wrong (chi=4) Jacobian.

Sibling to #700 (same #667/#670 direction-dependent CTM feature cluster, different commit `004fc89`/PR #671 and a different path). This one is `_ctm_tensor_moves.py` 2x2 absorption bond bookkeeping.

Surfaces as two `@pytest.mark.slow` failures in `tests/test_ctm_energy_implicit_chi_bump.py` (#692):
`test_ad_gradient_invariance_bump_vs_fixed_chi_max`, `test_ad_gradient_matches_fd_with_bump`.

## Evidence

Reproducer: `_build_site_tensor(D=2, d=2, seed=42)`, `heisenberg_gate()`, `max_iter=200, min_iter=2, conv_tol=1e-12, gmres_tol=1e-8`. Forced bump (`threshold=1e-12`, `step_size=2`, `chi_max=8`) vs `chi=8, increase_chi=False`.

Forward energies (fresh JAX cache, deterministic):

| path | `815bc76` (pre-#676) | `main` (post-#676) |
|---|---|---|
| bump 4→8 | 0.4837826747 | **0.4836777486** |
| fixed χ=8 | 0.4837826747 | **0.4837477082** |
| \|bump − fixed\| | **3.8e-14** ✓ | **7.0e-5** ✗ |

Note the **no-bump** fixed-χ=8 forward *also* shifted (0.4837826747 → 0.4837477082), so #676 changed the plain single-site 2x2 CTM fixed point for a direction-dependent (random `A.l != A.r`) tensor — the bump is not required to see the drift.

AD vs FD on the bump path (fresh cache, identical to polluted-cache run):
```
grad_ad[:5] = [-0.01578185 -0.04362361  0.00103752 -0.00012217  0.01822368]
grad_fd[:5] = [-0.26757489 -0.24392686 -0.35002565  0.16205265  0.01678625]
```
`grad_ad` is ~10× smaller than `grad_fd` on 27/32 indices. The bump fires deterministically on iter 1 (threshold=1e-12) for every FD probe, so `grad_fd` is not contaminated by bump-timing discontinuity — the AD gradient is genuinely wrong.

## Bisect

`git bisect` good `815bc76`, bad `main` (136 commits) → first bad `7b8e5ad` (PR #676). Checker: `|E_bump − E_fixed8| < 1e-9`. Good side reports 4e-14; bad side 7e-5.

## Fix notes

Any fix must preserve #676's direction-dependent bond bookkeeping intent while (a) restoring trajectory-independence of the single-site 2x2 CTM fixed point (bump 4→8 == fixed 8), and (b) restoring the chi-lock backward so implicit-AD gradients match FD. Candidate: verify the carried-vs-compressed bond pairing introduced in `_ctm_tensor_moves.py` is symmetric under the two convergence trajectories and that `f_bwd`'s chi bound uses `chi_post`, not the closure-captured `chi_initial`.

## Related

#700 (U(1)-Sz single-site env collapse, PR #671), #667/#670/#671/#676 (direction-dependent symmetric CTM). The two prototype guards are xfailed pending this issue via #692.

### PR DESCRIPTION
**Pull request by yingjerkao**
> 🤖 **AI-generated PR** — written by Claude Code, posted by @yingjerkao.

Fixes #702.

## Root cause

The 2x2-recipe sweep's **corner label convention was not sweep-invariant**. The env init (`_STD_EDGE_SPECS`) fixes which physical leg each corner label holds (with C4/T3/T4 names historically swapped vs geometry: `c4_r`=UP leg, `c4_u`=RIGHT leg). The left/top/right/bottom absorptions each *read* corners by label and *write* relabeled results — but the TOP move's C1 write, the RIGHT move's C2 write, and the BOTTOM move's C4+C3 writes relabeled the new corner with its two legs **swapped** relative to the init convention that `_build_enlarged_corner`, the move reads, and the energy/RDM contraction all assume.

- **Pre-#676** the defect was *uniform* — all four corners transposed once per sweep, at the same cycle point. The single-site scheme tolerated this global transposition (trajectory-independent, but converging to the transposed-scheme fixed point).
- **PR #676** made C3's read/write cycle self-consistent (needed for direction-dependent 2-site) but left C1/C2/C4 swapped — a *non-uniform* defect. The 1x1 fixed point became trajectory-dependent (#702: chi-bump vs fixed-chi drift 7e-5, implicit-AD gradient ~10x off, D=2 sharding parity blown to 7.7e-5).

The per-corner convention automaton (which physical leg each label holds, traced through the sweep order [left, top, right, bottom] from the init ground truth) pins every symptom and independently re-derives why #676's C3 change was correct for 2-site.

## Fix

14 line-targeted edits in `_ctm_tensor_absorb_{right,top,bottom}_2plaq` (+ their `_fused` fermionic twins): every corner write now lands in the natural init convention, and the RIGHT move's C3 read pairs `c3_l↔t3_l` (natural adjacency). #676's `bottom_left` enlarged-corner pairing **stays** — it *is* the natural adjacency given C4's historically swapped names — and is now correct for both the 1x1 and 2-site tilings because the convention no longer varies across the sweep.

Untouched (own consistent conventions or dead code): the legacy 1x1-recipe moves, `_ctm_tensor_paired_moves`, `_ctm_compiled_moves` (test-only), and the unused `_ctm_tensor_move_*_2x2` wrappers.

## Verification

| gate | before | after |
|---|---|---|
| bump 4→8 vs fixed χ=8 forward (seed 42) | 7.0e-5 ✗ | **2.2e-15** ✓ |
| same, seeds 0/1/3/7/11/13 | — | ≤ 1.9e-14 ✓ |
| `test_ad_gradient_matches_fd_with_bump` (un-xfailed) | AD ~10x off FD ✗ | **passes** ✓ |
| #670 2-site direction-dependent guards (5) | pass | **pass** ✓ |
| U(1)-Sz symmetric guards (12) | pass | **pass** ✓ |
| D=2 sharding parity (#706 xfail hook removed) | 7.68e-5 ✗ | **passes outright** ✓ |
| core suite + CTM test files | — | pass ✓ |

The single-site fixed point moves (seed-42 probe energy 0.4837826747 → 0.4843575437): the old value was the *transposed-scheme* fixed point; the energy/RDM contraction always assumed the natural convention, so it was reading transposed corners on 1x1.

## Collateral improvement: the #425/#426 random-init plateau is largely resolved

The historic "CTM plateaus at sv_diff~0.05 on random init while variPEPS converges in 9 iters" behavior (#425/#426, `project_tenax_ctm_doesnt_converge_random_init`) was another symptom of the non-invariant corner convention. Post-fix, the random-init D=2 χ=4 sweep converges to sv_diff ≈ 1e-6 (4 orders deeper), stalling only on the degenerate-pair SV basis rotation. All three previously-xfailed plateau tests in `test_ctm_python_loop.py` now **pass as genuine convergence guards** (`conv_tol=1e-5`, above the ~1e-6 floor; the narrowed xfail bands are kept for platforms whose BLAS still lands in the old regime): `test_python_loop_converges`, `test_python_loop_2site_checkerboard_converges`, `test_env_init_warm_start` — 13 passed, 0 xfailed.

## Known residual (documented, not a regression of this PR's scope)

Mutual bump-vs-fixed **gradient** agreement floors at ~1e-3 (pre-#676: 1e-13). The natural convention restores the double layer's ket-bra Z2 at the fixed point → **paired-degenerate projector SVs** (the #425 design-note mechanism); LAPACK's basis choice inside a degenerate pair differs between trajectories, and only the backward feels it. Both gradients are individually FD-consistent (probe: fixed-chi path shows the identical AD-vs-FD deviation profile as the bump path). The invariance gate is retuned to a norm-ratio check (0.9–1.1; the chi-lock breakage signature is a ~10x norm shrink) + `allclose(5e-2)`, with the measured floors documented in the test.

## Test changes

- `test_ctm_energy_implicit_chi_bump.py`: un-xfail the two #702-gated AD tests; add `test_forward_fixed_point_trajectory_independent` (forward-only #702 regression gate); retune the invariance asserts to the degenerate-pair floor.
- `test_ctm_sharding_parity.py`: remove the conditional #702 xfail hook so future parity misses fail loudly.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Adding the `run-full-tests` label rationale: this changes the CTM sweep's corner convention, which the required `-m core` subset only partially exercises. The slow-marked #702 gates (`test_ctm_energy_implicit_chi_bump.py`) and the #670 direction-dependent guards were run locally (all green — see PR body table), but a full-suite CI run is worth having before merge.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Merged `origin/main` in (was 2 commits behind; branch protection requires up-to-date). That pulled in **#707**, which added `test_ctm_recipe_2x2_consistency.py` and `test_ctm_recipe_2x2_production_correctness.py` — coverage for the same default 2x2 recipe this PR changes, so worth confirming explicitly: **all 3 pass** against the fix (`test_uniform_tiling_consistency[2x2]`, `[1x1]`, `test_bipartite_checkerboard_physical_energy`). They were written after the regression landed but don't encode the broken fixed point.

Full-suite matrix (9 jobs incl. the `slow` bucket) is now queued via the `run-full-tests` label, which covers the slow-marked #702 gates and #670 guards on CI rather than only locally.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

**The `Full tests (macos-latest, Python 3.12, fast-ipeps)` failure is pre-existing on `main`, not caused by this PR** — filed as #713.

`tests/test_ipeps.py::TestRDM::test_rdm_positive_semidefinite` fails with unbounded vertical-RDM eigenvalues `[-11.05971718, -0.59874749, 0.92557285, 11.73289182]`. The **identical four values** appear on `main` at `149366e` (the #708/#700 fix), which predates this branch. Bit-identical numerics across two different CTM code states means the failing path is insensitive to the corner-convention change here.

- ubuntu-latest 3.11 + 3.12 `fast-ipeps`: **pass**
- local Linux, this exact test on this branch: **passes**
- All three **required** checks — `Tests (Python 3.11)`, `Tests (Python 3.12)`, `Tests (macOS, Python 3.12)` — are **green**

It went unnoticed because `nightly.yml` runs tests on ubuntu only (macos-14 is used solely for wheel builds), and `Full tests (...)` is not a required check — so it only surfaces on PRs carrying `run-full-tests`, which is how it turned up here.

Re-enabling auto-merge: this PR does not introduce the failure, and blocking on it would mean blocking on a `main` defect.

File: src/tenax/algorithms/_ctm_tensor_moves.py [hunk 1]
```diff
@@ -664,7 +664,7 @@ def _ctm_tensor_absorb_top_2plaq(
 
     # ---- C1: project with P_top_left (≡ variPEPS .right of left-plaq) ----
     C1_new = _apply_proj_unfused(P_top_left, C1g, "c1_r", "l2")
-    C1_new = C1_new.relabels({"chi_new": "c1_d", "t4_u": "c1_r"})
+    C1_new = C1_new.relabels({"chi_new": "c1_r", "t4_u": "c1_d"})
```
[the remark this answers is not in the document]
Comment (yingjerkao) [reply src/tenax/algorithms/_ctm_tensor_moves.py:667]: > 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Confirmed live on `main`, and raised above its original P2 — now **#798**.

The write sites do leave the corners reversed (`C1_new.relabels({"chi_new": "c1_r", "t4_u": "c1_d"})` yields `("c1_r", "c1_d")`, not the canonical `("c1_d", "c1_r")`), and the consumer you named is purely positional — `_apply_sigma_to_corner` opens with `idx0, idx1 = corner.indices`. So `_sigma_gauge_fix_env(env.C1, s4, s1)` lands the left/down transform on `c1_r`.

Worth adding: `forward_gauge="sigma"` is known to blow up on the 2-site path, which is currently recorded as an unexplained property of the sigma gauge. A systematically mis-applied bond transform predicts exactly that shape — fine under the label-based default, divergent under the one path that reads corners positionally. Not asserting the causal link, but it is cheap to test and either fixes a feature or kills a hypothesis. The issue also proposes making the consumer label-based rather than only fixing the write order, since the class of defect is "a positional read of a label-based structure".
